src/s3path.jl

```diff
@@ -333,14 +333,57 @@ function FilePathsBase.sync(f::Function, src::AbstractPath, dst::S3Path; delete=
     end
 end
 
+
+# for some reason, sometimes we get back a `Pair`
+# other times a `AbstractDict`.
+function _pair_or_dict_get(p::Pair, k)
+    first(p) == k || return nothing
+    return last(p)
+end
+_pair_or_dict_get(d::AbstractDict, k) = get(d, k, nothing)
+
+function _readdir_add_results!(results, r, key_length)
+    rm_key = s -> chop(s, head=key_length, tail=0)
+    sizehint!(results, length(results) + parse(Int, r["KeyCount"]))
+    if haskey(r, "CommonPrefixes")
+        for p in r["CommonPrefixes"]
+            prefix = _pair_or_dict_get(p, "Prefix")
+            if prefix !== nothing
+                push!(results, rm_key(prefix))
+            end
+        end
+    end
+    if haskey(r, "Contents")
+        for o in r["Contents"]
+            prefix = _pair_or_dict_get(o, "Key")
+            if prefix !== nothing
+                push!(results, rm_key(prefix))
+            end
+        end
+    end
+    return get(r, "NextContinuationToken", nothing)
+end
+
 function Base.readdir(fp::S3Path; join=false, sort=true)
     if isdir(fp)
         k = fp.key
-        # Only list the files and "dirs" within this S3 "dir"
-        objects = s3_list_objects(fp.config, fp.bucket, k; delimiter="")
+        key_length = length(k)
 
-        # Only list the basename and not the full key
-        results = unique!([s3_get_name(k, string(o["Key"])) for o in objects])
+        results = String[]
+        r = @repeat 4 try
+            S3.list_objects_v2(fp.bucket, Dict("delimiter" => "/", "prefix" => k); aws_config=fp.config)
+        catch e
+            @delay_retry if ecode(e) in ["NoSuchBucket"] end
+        end
+        token = _readdir_add_results!(results, r, key_length)
```

Comment:
Let's just start `token` off in a way where we always enter the loop once:
```suggestion
        token = ""
```

Reply:
hm, is passing `continuation-token=""` the same as not passing it at all? that was my concern about something like that

Reply:
I get `ERROR: AWSException: InvalidArgument -- The continuation token provided is incorrect` when I try that locally. but I could unify them a bit more